p4_symbolic/ir/ir.proto

```diff
@@ -0,0 +1,230 @@
+// The IR does not cover all the features of P4 programs.
+// Specifically, we intentionally ignore parsers, de-parsers, stacks and unions,
+// checksum calculations, and extern definitions.
+// Additionally, some types of expressions and statements are unsupported, these
+// are described in the relevant sections of this file.
+// See the bmv2 JSON format reference for more information:
+// https://github.com/p4lang/behavioral-model/blob/master/docs/JSON_format.md
+
+syntax = "proto3";
+
+package p4_symbolic.ir;
+
+import "p4_pdpi/ir.proto";
+import "p4_symbolic/bmv2/bmv2.proto";
+
+// Overall structure of a p4 program.
+message P4Program {
+  // Header definitions, keyed by the header type name (e.g. standard_metadata).
+  map<string, HeaderType> headers = 1;
+  // Action definitions and implementations, keyed by action name.
+  map<string, Action> actions = 2;
+  // Table definitions and implementations, keyed by table name.
+  map<string, Table> tables = 3;
+  // The name of the initial table.
+  string initial_table = 6;
+}
+
+// A header definition. Includes the programmer assigned name,
+// the header id, and the fields in the heade.
+message HeaderType {
+  string name = 1;
+  int32 id = 2;
+  // Maps a field name to its definition.
+  map<string, HeaderField> fields = 3;
+}
+
+// A P4 header type definition.
+// Includes its name, bitwidth (type), and sign flag, and the
+// enclosing header type name that defines this field.
+message HeaderField {
+  string name = 1;
+  int32 bitwidth = 2;
+  bool signed = 3;
+  string header_type = 4;
+}
+
+// Overall action structure, combining definition with implementation.
+message Action {
+  // This contains the name, id, and parameters, among other things.
+  // Essentially, this is the action signature.
+  pdpi.ir.IrActionDefinition action_definition = 1;
+  // Action implementation, essentially its body.
+  ActionImplementation action_implementation = 2;
+}
+
+// Action implementation details.
+message ActionImplementation {
+  // List of statements in action body in sequence.
+  repeated Statement action_body = 1;
+}
+
+// Overall table structure, combining definition with implementation.
+message Table {
+  // This contains the name, id, action list, field matches, and max size.
+  pdpi.ir.IrTableDefinition table_definition = 1;
+  // This contains table implementation details.
+  TableImplementation table_implementation = 2;
+}
+
+// Table implementation details that are not exposed in the definition.
+message TableImplementation {
+  // The overal type of the match(s) of the table.
+  // In case the table have several keys with different match type,
+  // the overall match type of the table would be the most general.
+  // There is always a most general match because of P4 restrications
+  // on match combination (which may be under-documented).
+  // See https://github.com/p4lang/p4-spec/issues/411
+  p4_symbolic.bmv2.TableMatchTypeEnum match_type = 5;
```

Comment:
Do we need this? What will this be used for?

Reply:
I think it might help us with the SMT constraints: we can use it to figure out what the most general match is, specifically whether there is any LPM matches.

In some ways this information can be re-deduced from the individual matches. But since it exists, I thought I would keep it and use it directly.

Reply:
I left it this way because I am optimizing this for consumption rather than production. I thought a little redundancy might be ok if it can save us some code/computation at the consumers. There are other locations in this IR definition and also PDPI with some similar redundancy.